src/test/kotlin/com/atlassian/performance/tools/ssh/api/SshTest.kt

```diff
@@ -0,0 +1,69 @@
+package com.atlassian.performance.tools.ssh.api
+
+import org.junit.Assert
+import org.junit.Test
+import java.time.Duration
+import kotlin.system.measureTimeMillis
+
+class SshTest {
+
+    @Test
+    fun shouldDetachProcess() {
+        SshContainer().useSsh { sshHost ->
+            installPing(sshHost)
+
+            val ping = sshHost.newConnection().use { ssh ->
+                @Suppress("DEPRECATION") // tests public API
+                ssh.startProcess("ping localhost")
+            }
+            sshHost.newConnection().use { ssh ->
+                @Suppress("DEPRECATION") // tests public API
+                ssh.stopProcess(ping)
+            }
+        }
+    }
+
+    @Test
+    fun shouldNotWaitForBackground() {
+        SshContainer().useSsh { sshHost ->
+            val runMillis = measureTimeMillis {
+                sshHost.runInBackground("sleep 8")
+            }
+
+            Assert.assertTrue(runMillis < 1000)
+        }
+    }
+
+    @Test
+    fun shouldGetBackgroundResults() {
+        SshContainer().useSsh { sshHost ->
+            installPing(sshHost)
+
+            val ping = sshHost.runInBackground("ping localhost")
+            Thread.sleep(2000)
+            // meanwhile we can create and kill connections
+            sshHost.newConnection().use { it.safeExecute("ls") }
+            Thread.sleep(2000)
+            val pingResult = ping.stop(Duration.ofMillis(20))
+
+            Assert.assertTrue(pingResult.isSuccessful())
+            Assert.assertTrue(pingResult.output.contains("localhost ping statistics"))
+        }
+    }
+
+    @Test
+    fun shouldTolerateEarlyFinish() {
+        SshContainer().useSsh { sshHost ->
+            installPing(sshHost)
+
+            val fail = sshHost.runInBackground("nonexistent-command")
+            val failResult = fail.stop(Duration.ofMillis(20))
```

Comment:
This line flakes when sometimes it doesn't get the process' exit status. I'm not sure what's the condition.

```
java.lang.IllegalStateException: command.exitStatus must not be null
	at com.atlassian.performance.tools.ssh.WaitingCommand.waitForResult(WaitingCommand.kt:24)
	at com.atlassian.performance.tools.ssh.SshjBackgroundProcess.stop(SshjBackgroundProcess.kt:20)
	at com.atlassian.performance.tools.ssh.api.SshTest$shouldTolerateEarlyFinish$1.invoke(SshTest.kt:61)
	at com.atlassian.performance.tools.ssh.api.SshTest$shouldTolerateEarlyFinish$1.invoke(SshTest.kt:8)
	at com.atlassian.performance.tools.ssh.api.SshContainer.useSsh$ssh(SshContainer.kt:19)
	at com.atlassian.performance.tools.ssh.api.SshTest.shouldTolerateEarlyFinish(SshTest.kt:56)
```

I managed to make the problem frequently reproducible by wrapping the "when->then" part of the test in a `repeat(100) { ... }` block.

```
    @Test
    fun shouldTolerateEarlyFinish() {
        SshContainer().useSsh { sshHost ->
            installPing(sshHost)

            repeat(100) {
                val fail = sshHost.runInBackground("nonexistent-command")
                val failResult = fail.stop(Duration.ofMillis(20))

                Assert.assertEquals(127, failResult.exitStatus)
            }
        }
    }
```

Our current API of `BackgroundProcess#stop` (which is failing) says:
> Throws if getting the [SshConnection.SshResult] fails.

So the problem is not really a contract bug.

==== Start of my thoughts on the topic ====

I personally think that if it's possible to not handle exception then by default it shouldn't break your application, so it shouldn't really be thrown. Because of this I think this is an API design bug and IMHO we should fix it by introducing API that is returning nullable process result.

If we want to let the consumer distinguish reason of failures (missing result) we can do that by allowing the overwrite of problem handlers - by default returning `null`, but can throw if consumer wishes to do that. This way we don't leak the implementation specific exception layer into the interface, which should be polymorphic.

==== End of my thoughts on the topic ====

Nevertheless with the current API it's not a bug, yet it's a test failure. I'm trying to figure out how we can resolve it.

Possibly we could add `try`-`catch` block?

@dagguh do you remember if this test could work as expected if it would accept the process to not return exit status? Maybe it would be fine to repeat the `runInBackground` + `stop` couple of times? It doesn't sound as the best solution for me, but I struggle to find any better.

Reply:
The expectation is clear: `Assert.assertEquals(127, failResult.exitStatus)`. That's what happens in a shell for a non existent command. That doesn't happen - why? We shouldn't try to hide that dysfunction.

Reply:
Hypothesis 1: this would also flake for `SshConnection.execute`
Hypothesis 2: it's caused by the `BackgroundProcess.stop` timeout

Reply:
> Hypothesis 2: it's caused by the BackgroundProcess.stop timeout

Based on code in  https://github.com/atlassian/ssh/blob/9e6442e63fc4d17c7c994020be4d6f0fe30ace6d/src/main/kotlin/com/atlassian/performance/tools/ssh/WaitingCommand.kt#L29-L45 I believe that the timeout causes another exception to be thrown. I don't think it's rethrown or hidden in any way. on the call stack back to the `stop()`

Reply:
> I tried to add close and run the test with repeat(100), but it didn't help (unless I checked something wrong.

You can open a PR with that test.

Reply:
Yet this proves nothing really, it's only a clue

Reply:
Even with `repeat(1000)`

Reply:
https://github.com/atlassian/ssh/pull/23